### DISCUSSION TITLE
How to mock a specific method on an imported module?

### DISCUSSION THREAD

**Discussion by garyking** [vitest-dev/vitest#5100] [Q&A]
I have `a.ts` file:

```ts
import fsExtra from 'fs-extra';

export function doesPathExist(): boolean {
  return fsExtra.pathExistsSync(__filename);
}
```

I have `a.test.ts` file:

```ts
import mockFsExtra from 'fs-extra';
import { describe, test, expect, vi } from 'vitest';

import * as a from './a.js';

vi.mock('fs-extra', async () => {
  const fsExtra = await vi.importActual('fs-extra');

  return {
    ...fsExtra,

    pathExistsSync: vi.fn(),
  };
});

describe('a', () => {
  test('does path exist', () => {
    mockFsExtra.pathExistsSync.mockReturnValue(false);

    expect(a.doesPathExist()).toBe(false);
  });
});
```

Yet the test fails with the error `TypeError: default.pathExistsSync.mockReturnValue is not a function`. 

What am I doing wrong here? I am trying to mock fsExtra.pathExistsSync and set my own value, but it isn't working.

Thanks in advance!

**Comment by hi-ogawa** [comment]
As indicated by the error message, you probably need to swap out `fsExtra.default.pathExistsSync` instead of `fsExtra.pathExistsSync`.

https://stackblitz.com/edit/vitest-dev-vitest-jsdt7q?file=test%2Frepro.test.ts

```ts
vi.mock('fs-extra', async () => {
  const fsExtra = await vi.importActual('fs-extra');

  return {
    ...fsExtra,

    // -- not here
    // pathExistsSync: vi.fn()

    // ++ but here
    default: {
      ...fsExtra.default,
      pathExistsSync: vi.fn(),
    },
  };
});
```

**Reply by garyking** [reply]
Thanks, I'm really getting the hang of it. The only issue I have now is, how can I get the types to work in TypeScript? With the following code, the tests pass, but I have TypeScript errors, which I've indicated with inline comments: 

a.ts:

```ts
import fsExtra from 'fs-extra';
import { getDirnameAndFilename } from 'gary-utils/esm.js';

const { filename } = getDirnameAndFilename(import.meta.url);

export function doesPathExist(): boolean {
  const called = String(fsExtra.readFileSync(filename));

  console.log('called', called);

  return fsExtra.pathExistsSync(filename);
}
```

a.test.ts:

```ts
import fsExtra from 'fs-extra';
import { describe, test, expect, vi } from 'vitest';

import * as a from './a.js';

vi.mock('fs-extra', async (importOriginal) => {
  const original = await importOriginal<typeof fsExtra>();

  return {
    ...original,

    default: {
      // Property 'default' does not exist on type 'typeof
      // import("node_modules/@types/fs-extra/index")'.ts(2339)
      ...original.default,
      pathExistsSync: vi.fn(),
    },
  };
});

const readFileSyncSpy = vi.spyOn(fsExtra, 'readFileSync');

describe('a', () => {
  test('does path exist', () => {
    // Property 'mockReturnValue' does not exist on type '(path: string) =>
    // boolean'.ts(2339)
    fsExtra.pathExistsSync.mockReturnValue(false);

    expect(a.doesPathExist()).toBe(false);

    expect(readFileSyncSpy).toHaveBeenCalled();
  });
});
```

Thanks again!

**Reply by hi-ogawa** [reply]
Yeah, I see typescript error on stackblitz as well.

For the 2nd one, Vitest provides a typing utility exactly for this purpose https://vitest.dev/api/vi.html#vi-mocked

```ts
vi.mocked(fsExtra.pathExistsSync).mockReturnValue(false)
```

For the 1st type error with `original.default`, I'm not sure if Vitest is doing anything wrong. Maybe fs-extra's typing issue or simply typescript's limitation of cjs typing support. You can probably create some simpler wrapper to make it work like:

```ts
  // mimic actual cjs import typing
  type T_names = typeof import('fs-extra');
  type T_cjs = T_names & { default: T_names };
  const fsExtra = await vi.importActual<T_cjs>('fs-extra');
```

https://stackblitz.com/edit/vitest-dev-vitest-ubm2d4?file=test%2Frepro.test.ts

**Reply by garyking** [reply]
Thanks that helps a great deal! Here's what I've ended up with, with no TypeScript errors or failing tests:

```ts
import fsExtra from 'fs-extra';
import { describe, test, expect, vi } from 'vitest';

import * as a from './a.js';

const mockFsExtra = vi.mocked(fsExtra);
const readFileSyncSpy = vi.spyOn(fsExtra, 'readFileSync');

type CreateType<T> = T & { default: T };

vi.mock('fs-extra', async (importOriginal) => {
  const original = await importOriginal<CreateType<typeof fsExtra>>();

  return {
    ...original,

    default: {
      ...original.default,
      pathExistsSync: vi.fn(),
    },
  };
});

describe('a', () => {
  test('does path exist', () => {
    mockFsExtra.pathExistsSync.mockReturnValue(false);

    expect(a.doesPathExist()).toBe(false);
    expect(readFileSyncSpy).toHaveBeenCalled();
  });
});
```

**Reply by garyking** [reply]
@hi-ogawa I've got a question regarding `default` exports. How can I know if a module has `default` export or not, so I know in advance if I need to mock it directly? Since I can `import` a module even if it using `module.exports` rather than a `default` module. And `console.log(fsExtra.default)` returns `undefined`.

Thanks!

**Reply by sheremet-va** [reply]
This is how ESM works. You can find documentation about it on MDN for example: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import
